README.rst
Django-EasyMoney
================

An easy to use money field for Django.
Handles type conversions and arithmetic seamlessly.


Requirements
------------

Python 2.7 and 3.3+, Django 1.6+.


Installation
------------

::

    pip install django-easymoney


Overview
--------

.. code:: python

    from django.db import models
    from easymoney import MoneyField

    class MyModel(models.Model):
        price = MoneyField(default=3.14)
        other_price = MoneyField(default=1)


    obj = MyModel()
    print obj.price # -> $3.14

    # Money arithmetic
    obj.price + obj.other_price # $4.14

    # Mix with ints and floats
    obj.price + 1   # $4.14
    obj.price + 0.2 # $3.34

    # No partial cents or crazy floats
    obj.price / 3   # $1.05


Settings
--------

A primary use of easymoney is global currency defined in settings
with global formatting and precision. Here is how you do that:

.. code:: python

    # These are default settings, code and locales refer to ones used in babel library
    CURRENCY_CODE = 'USD'
    CURRENCY_LOCALE = 'en_US'
    CURRENCY_DECIMAL_PLACES = 2

    # This is optional, for cases when you want to use some fake currency
    CURRENCY_FORMAT = '# points'


Several currencies
------------------

One money field can't store different currencies, however, you can create different money classes and model fields for them:

.. code:: python

    from easymoney import Money, MoneyField

    class GameMoney(Money):
        # CODE = '...'
        # LOCALE = '...'
        DECIMAL_PLACES = 0
        FORMAT = '# points'

    class GameMoneyField(MoneyField):
        MONEY_CLASS = GameMoney



CHANGELOG
0.7.2
- better support Django migrations

0.7.1
- clearer error messages in binary ops TypeErrors (oTree-org)
- support Money hashing in Python 3 (oTree-org)

0.7
- support unary ops (Gregor Müllegger)
- support 3-argument pow()
- fixed Money.__repr__ in Python 3 (Gregor Müllegger)
- some optimizations

0.6
- support Django 1.8 and 1.9 (Andy Hayden and me)
- support Python 3.5 (Gregor Müllegger)
- support python 3 (Andy Hayden)
- improved formatting logic (Gregor Müllegger)
- added divmod and pow to Money (Gregor Müllegger)

0.5
- support subclassing to support several currencies simultaneously

0.4.1
- support Django 1.7 including migrations
- added CURRENCY_FORMAT setting

0.4
- added format support to Money
- added CURRENCY_DECIMAL_PLACES setting
- __str__ now correctly returns encoded string

0.3.6
- support pickling

0.3.5
- fixed Money stringification with non-ascii currency symbol

0.3.4
- get rid of MoneySelect and MoneyRadioSelect

0.3.3
- fixed Money repr encoding

0.3.2
- added MoneyRadioSelect widget
- fixed null values in MoneySelect

0.3.1
- fixed installation issue

0.3
- use babel for money formatting
- support choices
- use input type="number"
- fixed convertion to float and hashing of Money
- fixed copy/deepcopy of Money

0.2
- added simplistic widget
- support comparing money with non numbers
- fixed precision issue

0.1
Initial working verion


tests/urls.py
from __future__ import absolute_import
from django.conf.urls import patterns, include, url
from django.contrib import admin


admin.autodiscover()

urlpatterns = patterns('',
    (r'^admin/', include(admin.site.urls)),
)


tests/admin.py
from __future__ import absolute_import
from django.contrib import admin

from .models import Product, Option


admin.site.register(Product)
admin.site.register(Option)


tests/manage.py
from __future__ import absolute_import
#!/usr/bin/env python
import os
from os.path import abspath, dirname, split
import sys


project_dir, _ = split(dirname(abspath(__file__)))
sys.path.insert(0, project_dir)


if __name__ == "__main__":
    os.environ.setdefault("DJANGO_SETTINGS_MODULE", "tests.settings")

    from django.core.management import execute_from_command_line

    execute_from_command_line(sys.argv)


easymoney/models.py
from decimal import Decimal, ROUND_HALF_UP

from django.db import models

from .forms import MoneyInput


class Money(Decimal):
    def __new__(cls, amount):
        return Decimal.__new__(Money, _sanitize(amount))

    # Money is immutable
    def __copy__(self):
        return self

    def __deepcopy__(self, memo):
        return self

    def __float__(self):
        """Float representation."""
        return float(Decimal(self))

    def __str__(self):
        return '$%s' % Decimal(self) # TODO: use babel

    def __repr__(self):
        return 'Money(%s)' % self

    def __eq__(self, other):
        if isinstance(other, Money):
            return Decimal.__eq__(self, other)
        elif isinstance(other, (int, long, float, Decimal)):
            return Decimal.__eq__(self, _sanitize(other))
        else:
            return False

# Set up money arithmetic
def _to_decimal(amount):
    if isinstance(amount, Decimal):
        return amount
    elif isinstance(amount, float):
        return Decimal.from_float(amount)
    else:
        return Decimal(amount)

def _sanitize(amount):
    if isinstance(amount, Money):
        return amount
    return _to_decimal(amount).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)

def _make_method(name):
    method = getattr(Decimal, name)
    return lambda self, other, context=None: \
        Money(method(self, _to_decimal(other), context=context))

ops = 'add radd sub rsub mul rmul floordiv rfloordiv truediv rtruediv div rdiv mod rmod'
for op in ops.split():
    name = '__%s__' % op
    maker = make_compare if op in {'eq', 'ne'} else _make_method
    setattr(Money, name, maker(name))


class MoneyField(models.DecimalField):
    __metaclass__ = models.SubfieldBase

    # NOTE: we specify default value for max_digits for extra ease
    def __init__(self, verbose_name=None, name=None, max_digits=12, **kwargs):
        self.max_digits, self.decimal_places = max_digits, 2
        models.Field.__init__(self, verbose_name, name, **kwargs)

    def to_python(self, value):
        value = models.DecimalField.to_python(self, value)
        if value is None:
            return value

        return Money(value)

    def get_prep_value(self, value):
        if value is None:
            return None
        return Decimal(self.to_python(value))

    def formfield(self, **kwargs):
        defaults = {
            'widget': MoneyInput,
        }
        defaults.update(kwargs)
        return super(MoneyField, self).formfield(**defaults)

        # if self.choices:
        #     # Also copy-pasted from Field.formfield
        #     defaults = {
        #         'choices': self.choices,
        #         'coerce': self.coerce,
        #         'required': not self.blank,
        #         'label': capfirst(self.verbose_name),
        #         'help_text': self.help_text,
        #         'widget': forms.CheckboxSelectMultiple
        #     }
        #     defaults.update(kwargs)
        #     return TypedMultipleChoiceField(**defaults)
        # else:


easymoney.py
from __future__ import absolute_import

import sys
from decimal import Decimal, ROUND_HALF_UP

import babel.core
import babel.numbers
from django import forms
from django.db import models
from django.conf import settings
import six


__all__ = ['Money', 'MoneyField']


# Set up money arithmetic
def _to_decimal(amount):
    if isinstance(amount, Decimal):
        return amount
    elif isinstance(amount, float):
        return Decimal.from_float(amount)
    else:
        return Decimal(amount)


def _make_unary_operator(name):
    method = getattr(Decimal, name, None)
    # NOTE: current context would be used anyway, so we can just ignore it.
    #       Newer pythons don't have that, keeping this for compatability.
    return lambda self, context=None: self.__class__(method(self))


def _prepare_operand(self, other):
    try:
        return _to_decimal(other)
    except:
        raise TypeError(
            "Cannot do arithmetic operation between "
            "{} and {}.".format(repr(self), repr(other))
        )


def _make_binary_operator(name):
    method = getattr(Decimal, name, None)
    def binary_function(self, other, context=None):
        other = _prepare_operand(self, other)
        return self.__class__(method(self, other))
    return binary_function


def format_currency(number, currency, format, locale=babel.numbers.LC_NUMERIC,
                    force_frac=None, format_type='standard'):
    """Same as ``babel.numbers.format_currency``, but has ``force_frac``
    argument instead of ``currency_digits``.

    If the ``force_frac`` argument is given, the argument is passed down to
    ``pattern.apply``.
    """
    locale = babel.core.Locale.parse(locale)
    if format:
        pattern = babel.numbers.parse_pattern(format)
    else:
        try:
            pattern = locale.currency_formats[format_type]
        except KeyError:
            raise babel.numbers.UnknownCurrencyFormatError(
                "%r is not a known currency format type" % format_type)
    if force_frac is None:
        fractions = babel.core.get_global('currency_fractions')
        try:
            digits = fractions[currency][0]
        except KeyError:
            digits = fractions['DEFAULT'][0]
        frac = (digits, digits)
    else:
        frac = force_frac
    return pattern.apply(number, locale, currency=currency, force_frac=frac)


# Data class

class Money(Decimal):
    # Class settings to override in descendants
    CODE = getattr(settings, 'CURRENCY_CODE', 'USD')
    FORMAT = getattr(settings, 'CURRENCY_FORMAT', None)
    LOCALE = getattr(settings, 'CURRENCY_LOCALE', 'en_US')
    DECIMAL_PLACES = getattr(settings, 'CURRENCY_DECIMAL_PLACES', 2)
    MIN_DECIMAL_PLACES = 2

    def __new__(cls, amount):
        return Decimal.__new__(cls, cls._sanitize(amount))

    @classmethod
    def _sanitize(cls, amount):
        if isinstance(amount, cls):
            return amount
        quant = Decimal('0.1') ** cls.DECIMAL_PLACES
        return _to_decimal(amount).quantize(quant, rounding=ROUND_HALF_UP)

    # Support for pickling
    def __reduce__(self):
        return (self.__class__, (Decimal.__str__(self),))

    # Money is immutable
    def __copy__(self):
        return self

    def __deepcopy__(self, memo):
        return self

    def __float__(self):
        """Float representation."""
        return float(Decimal(self))

    def __unicode__(self):
        return self._format_currency(Decimal(self))

    def __str__(self):
        string = self._format_currency(Decimal(self))
        if six.PY2:
            return string.encode('utf-8')
        return string

    @classmethod
    def _format_currency(cls, number):
        return format_currency(
            number=number,
            currency=cls.CODE,
            format=cls.FORMAT,
            locale=cls.LOCALE,
            force_frac=(cls.MIN_DECIMAL_PLACES, cls.DECIMAL_PLACES)
        )

    def __format__(self, format_spec):
        if format_spec in {'', 's'}:
            formatted = six.text_type(self)
        else:
            formatted = format(Decimal(self), format_spec)

        if isinstance(format_spec, six.binary_type):
            return formatted.encode('utf-8')
        else:
            return formatted

    def __repr__(self):
        return '{}({})'.format(self.__class__.__name__, Decimal.__str__(self))

    def __eq__(self, other):
        if isinstance(other, Money):
            return Decimal.__eq__(self, other)
        elif isinstance(other, six.integer_types + (float, Decimal)):
            return Decimal.__eq__(self, self._sanitize(other))
        else:
            return False

    # for Python 3:
    # need to re-define __hash__ because we defined __eq__ above
    # https://docs.python.org/3.5/reference/datamodel.html#object.%5F%5Fhash%5F%5F
    __hash__ = Decimal.__hash__

    # Special casing this, cause it have extra modulo arg
    def __pow__(self, other, modulo=None):
        other = _prepare_operand(self, other)
        return self.__class__(Decimal.__pow__(self, other, modulo))

    __abs__ = _make_unary_operator('__abs__')
    __pos__ = _make_unary_operator('__pos__')
    __neg__ = _make_unary_operator('__neg__')

    __add__ = _make_binary_operator('__add__')
    __radd__ = _make_binary_operator('__radd__')
    __sub__ = _make_binary_operator('__sub__')
    __rsub__ = _make_binary_operator('__rsub__')
    __mul__ = _make_binary_operator('__mul__')
    __rmul__ = _make_binary_operator('__rmul__')
    __floordiv__ = _make_binary_operator('__floordiv__')
    __rfloordiv__ = _make_binary_operator('__rfloordiv__')
    __truediv__ = _make_binary_operator('__truediv__')
    __rtruediv__ = _make_binary_operator('__rtruediv__')
    if hasattr(Decimal, '__div__'):
        __div__ = _make_binary_operator('__div__')
        __rdiv__ = _make_binary_operator('__rdiv__')
    __mod__ = _make_binary_operator('__mod__')
    __rmod__ = _make_binary_operator('__rmod__')
    __divmod__ = _make_binary_operator('__divmod__')
    __rdivmod__ = _make_binary_operator('__rdivmod__')
    __rpow__ = _make_binary_operator('__rpow__')

    def deconstruct(self):
        return '{}.{}'.format(self.__module__, self.__class__.__name__), \
               [Decimal.__str__(self)], {}


# Model field

class MoneyField(six.with_metaclass(models.SubfieldBase, models.DecimalField)):
    MONEY_CLASS = Money

    # NOTE: we specify default value for max_digits for extra ease
    def __init__(self, verbose_name=None, name=None, max_digits=12, **kwargs):
        self.max_digits, self.decimal_places = max_digits, self.MONEY_CLASS.DECIMAL_PLACES
        models.Field.__init__(self, verbose_name, name, **kwargs)

    def deconstruct(self):
        name, path, args, kwargs = super(MoneyField, self).deconstruct()
        del kwargs["decimal_places"]
        return name, path, args, kwargs

    def to_python(self, value):
        value = models.DecimalField.to_python(self, value)
        if value is None:
            return value

        return self.MONEY_CLASS(value)

    def get_prep_value(self, value):
        if value is None:
            return None
        return Decimal(self.to_python(value))

    def formfield(self, **kwargs):
        defaults = {
            'form_class': MoneyFormField,
            'choices_form_class': MoneyChoiceField,
        }
        defaults.update(kwargs)
        return super(MoneyField, self).formfield(**defaults)


# Form fields

class MoneyFormField(forms.DecimalField):
    def prepare_value(self, value):
        return to_dec(value)


class MoneyChoiceField(forms.TypedChoiceField):
    def prepare_value(self, value):
        return to_dec(value)

    def __init__(self, *args, **kwargs):
        super(MoneyChoiceField, self).__init__(*args, **kwargs)
        self.choices = [(to_dec(k), v) for k, v in self.choices]


# Utils

def to_dec(value):
    return Decimal(value) if isinstance(value, Money) else value

def stdout_encode(u, default='UTF8'):
    if sys.stdout.encoding:
        return u.encode(sys.stdout.encoding)
    return u.encode(default)
